AmplifyPlugins/Auth/Tests/AWSCognitoAuthPluginUnitTests/hierarchical-state-machine-swiftTests/StateMachineListenerTests.swift

```diff
@@ -17,97 +17,75 @@ class StateMachineListenerTests: XCTestCase {
         stateMachine = CounterStateMachine.logging()
     }
 
-    func testNotifiesOnListen() {
-        stateMachine.send(Counter.Event(id: "test", eventType: .increment))
-        let subscribed = expectation(description: "subscribed")
+    func testNotifiesOnListen() async {
```

Comment:
As noted above, we need another test that asserts the order of execution:

* **Given** a state machine in state S<sub>0</sub> at T<sub>0</sub>
* **When** 
    * I subscribe to the state machine at time T<sub>1</sub>, and
    * At time T<sub>2</sub>, another process sends an event that triggers a transition to state S<sub>1</sub>
* **Then**
    * My listener is invoked at time T<sub>i</sub> > T<sub>1</sub> with the state S<sub>0</sub>; and subsequently
    * My listener is invoked at time T<sub>j</sub> > T<sub>i</sub> with the state S<sub>1</sub>

Note that there is no assertion about the truth of _T<sub>2</sub> <= T<sub>i</sub>_

Since this is async, we may need to test this in a loop to gain confidence that we don't have a hard-to-reproduce race condition. What we want to simulate is basically that Thread1 starts a subscription at almost exactly the same time as Thread2 sends the event. The State Machine must serialize the processing of messages such that the subscription listener is invoked prior to processing the state transition. In the current code, that assertion may fail since we invoke the initial listener asynchronously.